python/ml4ir/base/model/scoring/monte_carlo_scorer.py

```diff
@@ -99,12 +91,171 @@ def __init__(
+        """
+        Compute scores for input features with fixed masks applied.
+
+        Parameters
+        ----------
+        inputs : dict of tf.Tensor
+            Dictionary containing input feature tensors. Each key represents the feature name, and each value is a tensor of shape [batch_size, feature_length].
+        masked_inputs_list : list of dict of tf.Tensor
+            A list where each element is a dictionary of masked input feature tensors. Each dictionary has the same keys as the input dictionary, and the values are tensors of the same shape, but with masks applied.
+        training : bool
+            Indicator of whether the model is in training mode. This parameter is passed to the super().call method.
+
+        Returns
+        -------
+        all_scores : tf.Tensor
+            Tensor containing the computed scores for all masked inputs. The shape of the tensor depends on the specific model's output.
+        """
+        # Stack the masked inputs to create a batch of different masked inputs
+        stacked_inputs = {key: tf.stack([masked_inputs[key] for masked_inputs in masked_inputs_list])
+                          for key in inputs.keys()}
+
+        # Flatten the batch dimension with the mask dimension
+        flattened_inputs = {}
+        for key, value in stacked_inputs.items():
+            value_shape = tf.shape(value)
+            flattened_shape = tf.concat([[value_shape[0] * value_shape[1]], value_shape[2:]], axis=0)
+            flattened_inputs[key] = tf.reshape(value, flattened_shape)
+
+        # Compute scores for all masked inputs in one go
+        all_scores = super().call(flattened_inputs, training=False)[self.output_name]
+        return all_scores
+
+    def reshape_and_normalize(self, all_scores, mask_count, batch_size):
+        """
+        Reshape the scores tensor to separate mask and batch dimensions, then normalize the scores.
+
+        Parameters
+        ----------
+        all_scores : tf.Tensor
+            Tensor containing scores computed for all masked inputs. The shape is expected to be [mask_count * batch_size, ...].
+        mask_count : int
+            The number of different masks applied. This corresponds to the number of Monte Carlo trials.
+        batch_size : int
+            The size of the batch of inputs.
+
+        Returns
+        -------
+        tf.Tensor
+            The reshaped and normalized scores tensor. The shape of the tensor will be [batch_size, ...].
+        """
+        all_scores_shape = tf.shape(all_scores)
+        all_scores = tf.reshape(all_scores, tf.concat([[mask_count, batch_size], all_scores_shape[1:]], axis=0))
+
+        # Aggregate the scores across the mask dimension
+        all_scores = tf.reduce_sum(all_scores, axis=0)
+        all_scores = tf.divide(all_scores, self.monte_carlo_fixed_trials_count)
+        return all_scores
+
+    def deterministic_call(self, inputs: Dict[str, tf.Tensor], training=None):
```

Comment:
@mohazahran I would recommend both stochastic and deterministic functions have the same function definition (args)